unic/utils/src/char_property.rs

```diff
@@ -0,0 +1,81 @@
+// Copyright 2017 The UNIC Project Developers.
+//
+// See the COPYRIGHT file at the top-level directory of this distribution.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+
+//! Taxonomy and Contracts for Character Property types.
+//!
+//! See also the list of types of character properties defined in the UCD:
+//! <http://unicode.org/reports/tr44/#About_Property_Table>, in [Unicode® Standard Annex #44 —
+//! Unicode Character Database](http://unicode.org/reports/tr44/#About_Property_Table)
+
+
+use std::fmt::{Debug, Display};
+use std::hash::Hash;
+
+
+// == Property Domain ==
+
+/// A Character Property defined for some characters.
+///
+/// Examples: *Decomposition_Type*, *Numeric_Type*
+pub trait PartialCharProperty: Copy + Debug + Display + Eq + Hash {
+    /// Find the character property value, or None.
+    fn of(ch: char) -> Option<Self>;
+}
+
+
+/// A Character Property defined on all characters.
+///
+/// Examples: *Age*, *Name*, *General_Category*, *Bidi_Class*
+pub trait CompleteCharProperty: PartialCharProperty + Default {
```

Comment:
This works very well. Thanks for the idea, @CAD97!

The only issue is the error if none of these are implemented and a Property Range type is used, in which case rustc claims that `CompleteCharProperty` is required to be implemented, which is invalid, and only `PartialCharProperty` is required. I suppose I should file an issue for this upstream.